ports/silabs/common-hal/busio/I2C.c

```diff
@@ -0,0 +1,212 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "shared-bindings/busio/I2C.h"
+#include "py/mperrno.h"
+#include "py/runtime.h"
+#include "shared-bindings/microcontroller/__init__.h"
+#include "supervisor/shared/translate/translate.h"
+#include "shared-bindings/microcontroller/Pin.h"
+
+STATIC I2CSPM_Init_TypeDef i2cspm_init;
+STATIC bool in_used = false;
+STATIC bool never_reset = false;
+
+// Reser I2C peripheral
+void i2c_reset(void) {
+    if ((!never_reset) && in_used) {
+        I2C_Reset(DEFAULT_I2C_PERIPHERAL);
+        in_used = false;
+    }
+}
+
+// Construct I2C protocol, this function init i2c peripheral
+void common_hal_busio_i2c_construct(busio_i2c_obj_t *self,
+    const mcu_pin_obj_t *scl,
+    const mcu_pin_obj_t *sda,
+    uint32_t frequency, uint32_t timeout) {
+
+    if ((scl != NULL) && (sda != NULL)) {
+        if (scl->function_list[ DEFAULT_I2C_PERIPHERAL == I2C1?
+                                FN_I2C1_SCL : FN_I2C0_SCL] == 1 &&
+            scl->function_list[DEFAULT_I2C_PERIPHERAL == I2C1?
+                               FN_I2C1_SDA : FN_I2C0_SDA] == 1) {
+            self->scl = scl;
+            self->sda = sda;
+            self->has_lock = false;
+            i2cspm_init.sclPort = self->scl->port;
+            i2cspm_init.sclPin = self->scl->number;
+            i2cspm_init.sdaPort = self->sda->port;
+            i2cspm_init.sdaPin = self->sda->number;
+            i2cspm_init.port = DEFAULT_I2C_PERIPHERAL;
+            i2cspm_init.i2cRefFreq = 0;
+            i2cspm_init.i2cMaxFreq = I2C_FREQ_STANDARD_MAX;
+            i2cspm_init.i2cClhr = i2cClockHLRStandard;
+
+            self->i2cspm = i2cspm_init.port;
+            I2CSPM_Init(&i2cspm_init);
+            common_hal_mcu_pin_claim(scl);
+            common_hal_mcu_pin_claim(sda);
+            in_used = true;
+        } else {
+            mp_raise_ValueError(translate("I2C busy pin"));
```

Comment:
The "free pin" validation in shared-bindings should be checking for this.

(Part of the point here is not to create new error messages if you can use existing ones, since it increases the size of the port and introduces new messages that have to be translated.)